File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    let depth = triangle.length
    // let answer = 0
    // function dfs(sum, count, position){
    //     if(count < depth){
    //         dfs(sum+triangle[count][position], count+1, Math.max(position, 0))
    //         dfs(sum+triangle[count][position], count+1, Math.min(position+1, count+1))
    //     } else answer = Math.max(answer, sum)
    // }
    // dfs(0, 0, 0)
    // return answer;
    triangle.map((el, idx) => {
        if(idx === 0) return
        for(let i=0; i<=idx; i++){
            el[i] += Math.max(triangle[idx-1][Math.max(0,i-1)], triangle[idx-1][Math.min(i, triangle[idx-1].length-1)])
        }
    })
    let answer = triangle[triangle.length-1].reduce((acc, cur) => {
        if(cur > acc) acc = cur
        return acc
    }, 0)
    return answer
}

File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    l = len(triangle)
    for i in range(l):
        for j, _ in enumerate(triangle[i]):
            if i == 0 and j == 0: continue
            elif j == 0: triangle[i][j] += triangle[i-1][j]
            elif j == i: triangle[i][j] += triangle[i-1][j-1]
            else: triangle[i][j] += max(triangle[i-1][j-1], triangle[i-1][j])

    return max(triangle[l-1])